pkg/provider/core.go

```diff
@@ -398,6 +400,32 @@ func decodeProviderSpec(machineClass *v1alpha1.MachineClass) (*api.EquinixMetalP
 	return providerSpec, nil
 }
 
+func createDeviceWithReservations(svc packngo.DeviceService, createRequest *packngo.DeviceCreateRequest, reservationIDs []string, reservedOnly bool) (device *packngo.Device, err error) {
+	// if there were no reservation IDs and I didn't ask for reservedOnly, then just create one on-demand and return
+	if len(reservationIDs) == 0 && !reservedOnly {
+		device, _, err = svc.Create(createRequest)
+		return device, err
+	}
+
+	// if we got here, we either had some reservation IDs, or we were asked to do reserved only.
+	// In both cases, we try reservations first.
+	for _, resID := range reservationIDs {
```

Comment:
Is this the only way to check which reservation ID can be used? It might be quite costly/stressing the EQXM API. Can't you list the devices and filter the in-use IDs from the `reservationIDs` list?

Reply:
There is an even better way, but that is contained entirely within `createDeviceWithReservations()`, which is why I created the separate function.

This still works (however inefficiently for many devices). I want to get this in, then open a separate PR just for that. keep this clean.

Reply:
OK